File: src/gui/gui.c
#include <stdarg.h>

#include <SDL2/SDL_ttf.h>

#include "gui.h"

#include "../util/util.h"
#include "../renderer/renderer.h"

typedef struct FontCacheEntry {
  TTF_Font *font;
  i32 size;
} FontCacheEntry;

static FontCacheEntry *fontCache;
static i64 fontCacheLength;
static i64 fontCacheCapacity;

TTF_Font *LoadFont(i32 size) {
  // Find the font with the required size in the loaded font cache.
  for (i64 i = 0; i < fontCacheLength; i++) {
    const auto entry = &fontCache[i];
    if (entry->size == size) {
      return entry->font;
    }
  }

  // Font not found. Ensure there is enough capacity to add it to the cache.
  if (fontCacheLength >= fontCacheCapacity) {
    if (fontCacheCapacity == 0) {
      fontCacheCapacity = 8;
    } else {
      fontCacheCapacity *= 2;
    }
    fontCache = Erealloc(fontCache, fontCacheCapacity * sizeof(fontCache[0]));
  }

  // Reserve the new font entry.
  auto entry = &fontCache[fontCacheLength];
  fontCacheLength++;

  // Open the font.
  auto font = TTF_OpenFont("assets/fonts/OpenSans-Bold.ttf", size);
  if (font == nullptr) {
    Fatalf("%s", SDL_GetError());
  }

  // Add the new font to the font cache.
  entry->font = font;
  entry->size = size;

  return font;
}

void RenderProgressBar(i32 x,
                       i32 y,
                       i32 width,
                       i32 height,
                       i32 strokeSize,
                       i32 progress) {
  RenderRect(x, y, width, strokeSize, 0, 0, 0);
  RenderRect(x, y + height - strokeSize - 1, width, strokeSize, 0, 0, 0);
  RenderRect(x, y, strokeSize, height, 0, 0, 0);
  RenderRect(x + width - strokeSize - 1, y, strokeSize, height, 0, 0, 0);

  const auto cropX = x + strokeSize;
  const auto cropY = y + strokeSize;
  const auto cropWidth = width - strokeSize * 2;
  const auto cropHeight = height - strokeSize * 2;

  u8 red = 0xFF;
  u8 green = 0x00;
  u8 blue = 0x00;

  if (progress > 75) {
    red = 0x00;
    green = 0xFF;
    blue = 0x00;
  } else if (progress > 30) {
    red = 0xFF;
    green = 0xFF;
    blue = 0x00;
  }

  RenderRect(
    cropX, cropY, cropWidth * progress / 100, cropHeight, red, green, blue);
}

void RenderString(i32 x,
                  i32 y,
                  i32 size,
                  u8 red,
                  u8 green,
                  u8 blue,
                  u8 alpha,
                  u32 flags,
                  i32 lineNumber,
                  const char *const format,
                  ...) {
  char text[4096];
  va_list args;
  va_start(args, format);
  vsnprintf(text, sizeof(text), format, args);
  va_end(args);

  if (text[0] == '\0') {
    return;
  }

  const SDL_Color color = {
    .r = red,
    .g = green,
    .b = blue,
    .a = alpha,
  };

  const auto font = LoadFont(size);

  auto fontSurface = TTF_RenderText_Blended(font, text, color);
  if (fontSurface == nullptr) {
    Fatalf("%s", SDL_GetError());
  }

  auto texture = SDL_CreateTextureFromSurface(renderer, fontSurface);
  if (flags == 1) {
    x = (WindowWidth - fontSurface->w) / 2;
    y = (WindowHeight + (fontSurface->h + 50) * lineNumber) / 2;
  }

  const SDL_Rect source = {
    .x = 0,
    .y = 0,
    .w = fontSurface->w,
    .h = fontSurface->h,
  };

  const SDL_Rect destination = {
    .x = x,
    .y = y,
    .w = fontSurface->w,
    .h = fontSurface->h,
  };

  SDL_RenderCopy(renderer, texture, &source, &destination);
  SDL_DestroyTexture(texture);
  SDL_FreeSurface(fontSurface);
}

void RenderHealthBar(i32 health) {
  constexpr i32 scale = 2;

  constexpr i32 x = 20;
  constexpr i32 y = 20;

  i32 cropX = 364;
  constexpr i32 cropY = 199;
  i32 cropWidth = 130;
  constexpr i32 cropHeight = 14;

  RenderSprite(0,
               x,
               y,
               cropWidth * scale,
               cropHeight * scale,
               cropX,
               cropY,
               cropWidth,
               cropHeight);

  cropX = 494;
  cropWidth = 7;

  if (health < 0) {
    return;
  }

  constexpr i32 maxElements = 18;

  auto elementCount = health * maxElements / 100;
  if (elementCount > maxElements) {
    elementCount = maxElements;
  }

  for (i32 i = 0; i < elementCount; i++) {
    RenderSprite(0,
                 x + i * cropWidth * scale,
                 y,
                 cropWidth * scale,
                 cropHeight * scale,
                 cropX,
                 cropY,
                 cropWidth,
                 cropHeight);
  }
}

void RenderSmallLogo(void) {
  constexpr i32 cropWidth = 115;
  constexpr i32 cropHeight = 58;

  constexpr i32 scale = 1;

  constexpr auto width = cropWidth * scale;
  constexpr auto height = cropHeight * scale;

  RenderSprite(
    0, WindowWidth - width, 0, width, height, 700, 401, cropWidth, cropHeight);
}

File: gui.go
package main

import (
	"fmt"

	"github.com/veandco/go-sdl2/sdl"
	"github.com/veandco/go-sdl2/ttf"
)

type FontCacheEntry struct {
	font *ttf.Font
	size int
}

var fontCache []FontCacheEntry

func loadFont(size int) *ttf.Font {
	// Find the font with the required size in the loaded font cache
	for _, entry := range fontCache {
		if entry.size == size {
			return entry.font
		}
	}

	font, err := ttf.OpenFont("assets/fonts/OpenSans-Bold.ttf", size)
	if err != nil {
		fatalf("can't open font file: %v", sdl.GetError())
	}

	// Add the new font to the font cache
	fontCache = append(fontCache, FontCacheEntry{font: font, size: size})

	return font
}

func renderProgressBar(rect sdl.Rect, strokeSize, progress int) {
	borderColor := sdl.Color{R: 0, G: 0, B: 0, A: 0}

	renderRect(sdl.Rect{X: rect.X, Y: rect.Y, W: rect.W, H: int32(strokeSize)}, borderColor)
	renderRect(sdl.Rect{X: rect.X, Y: rect.Y + rect.H - int32(strokeSize) - 1, W: rect.W, H: int32(strokeSize)}, borderColor)
	renderRect(sdl.Rect{X: rect.X, Y: rect.Y, W: int32(strokeSize), H: rect.H}, borderColor)
	renderRect(sdl.Rect{X: rect.X + rect.W - int32(strokeSize) - 1, Y: rect.Y, W: int32(strokeSize), H: rect.H}, borderColor)

	barColor := sdl.Color{R: 255, G: 0, B: 0, A: 255}
	switch {
	case progress > 75:
		barColor = sdl.Color{R: 0, G: 255, B: 0, A: 255}
	case progress > 30:
		barColor = sdl.Color{R: 255, G: 255, B: 0, A: 255}
	}

	barRect := sdl.Rect{
		X: rect.X + int32(strokeSize),
		Y: rect.Y + int32(strokeSize),
		W: (rect.W - int32(strokeSize*2)) * int32(progress) / 100,
		H: rect.H - int32(strokeSize*2),
	}

	renderRect(barRect, barColor)
}

func renderString(
	x, y, size int,
	color sdl.Color,
	relativeToWindowCenter bool,
	lineNo int,
	format string,
	args ...any,
) {
	text := fmt.Sprintf(format, args...)

	if len(text) == 0 {
		return
	}

	font := loadFont(size)

	fontSurface, err := font.RenderUTF8Blended(text, color)
	if err != nil {
		fatalf("can't get font surface: %v", err)
	}
	defer fontSurface.Free()

	texture, err := renderer.CreateTextureFromSurface(fontSurface)
	if err != nil {
		fatalf("can't create texture: %v", err)
	}
	defer texture.Destroy()

	src := sdl.Rect{X: 0, Y: 0, W: fontSurface.W, H: fontSurface.H}

	dest := sdl.Rect{X: int32(x), Y: int32(y), W: fontSurface.W, H: fontSurface.H}
	if relativeToWindowCenter {
		dest.X = (WindowWidth - fontSurface.W) / 2
		dest.Y = (WindowHeight + (fontSurface.H+50)*int32(lineNo)) / 2
	}

	renderer.Copy(texture, &src, &dest)
}

func renderHealthBar(health int) {
	const scale = 2

	const (
		x = 20
		y = 20
	)

	crop := sdl.Rect{X: 364, Y: 199, W: 130, H: 14}

	renderSprite(0, sdl.Rect{X: x, Y: y, W: crop.W * scale, H: crop.H * scale}, crop)

	crop.X = 494
	crop.W = 7

	if health < 0 {
		return
	}

	const maxElems = 18
	elemNum := min(maxElems, health*maxElems/100)

	for i := range elemNum {
		renderSprite(0, sdl.Rect{X: x + int32(i)*crop.W*scale, Y: y, W: crop.W * scale, H: crop.H * scale}, crop)
	}
}

func renderSmallLogo() {
	const scale = 1

	const (
		cropW = 115
		cropH = 58
	)

	w := cropW * scale
	h := cropH * scale

	renderSprite(0, sdl.Rect{X: int32(WindowWidth - w), Y: 0, W: int32(w), H: int32(h)}, sdl.Rect{X: 700, Y: 401, W: cropW, H: cropH})
}
